font/src/ft/mod.rs

```diff
@@ -537,48 +607,66 @@ impl FreeTypeRasterizer {
-                if let (Some(path), Some(_)) = (pattern.file(0), pattern.index().next()) {
-                    match self.keys.get(&path) {
-                        // We've previously loaded this font, so don't
-                        // load it again.
-                        Some(&key) => {
-                            debug!("Hit for font {:?}; no need to load", path);
-                            // Update fixup factor
-                            self.faces.get_mut(&key).unwrap().pixelsize_fixup_factor =
-                                pattern.pixelsizefixupfactor().next();
-                            Ok(key)
-                        },
-
-                        None => {
-                            debug!("Miss for font {:?}; loading now", path);
-                            // Safe to unwrap the option since we've already checked for the path
-                            // and index above.
-                            let key = self.face_from_pattern(&pattern)?.unwrap();
-                            Ok(key)
-                        },
+        // This check should be pretty fast and save us from traversing the entire list
+        if !fallback_list.coverage.has_char(ch) {
+            return Ok(glyph.font_key);
+        }
+
+        for font_pattern in &fallback_list.list {
+            let path = match font_pattern.file(0) {
+                Some(path) => path,
+                None => continue,
+            };
+
+            match self.keys.get(&path) {
+                Some(&key) => {
+                    let face = match self.faces.get(&key) {
+                        Some(face) => face,
+                        None => continue,
+                    };
+
+                    let index = face.ft_face.get_char_index(ch as usize);
+
+                    // We found something in a current face, so let's use it
+                    if index != 0 {
+                        return Ok(key);
                     }
-                } else {
-                    Err(Error::MissingFont(FontDesc::new(
-                        "fallback-without-path",
-                        Style::Specific(glyph.to_string()),
-                    )))
-                }
-            },
-            None => Err(Error::MissingFont(FontDesc::new(
-                "no-fallback-for",
-                Style::Specific(glyph.to_string()),
-            ))),
+                },
+                None => {
+                    // This path will be hit, only if you don't have glyph at all or didn't
+                    // loaded font, so not that hot for a normal scenario and all costly parts
+                    // are inside the if
+                    let pattern = font_pattern.clone();
+                    let charset = pattern.get_charset();
+
+                    if charset.has_char(ch) {
+                        let config = fc::Config::get_current();
+
+                        // Recreate a pattern
+                        let mut pattern = fc::Pattern::new();
```

Comment:
Why do we have to create a new pattern here, instead of using the `pattern` we cloned from `font_pattern` and then just setting the pixelsize on that?

Reply:
This is a specific thing to docs... You should use your "original pattern" (the one you've used to font_sort) and use a thing from font_sort(the font_pattern) + render_pattern to get a valid thing.